File: app/src/main/java/org/amoustakos/boilerplate/examples/ui/presenters/ActivityListingPresenter.java
package org.amoustakos.boilerplate.examples.ui.presenters;

import android.content.pm.PackageManager;

import org.amoustakos.boilerplate.examples.io.local.models.ActivityListingModel;
import org.amoustakos.boilerplate.examples.ui.contracts.ActivityListingContract;
import org.amoustakos.boilerplate.ui.activities.BaseActivity;
import org.amoustakos.boilerplate.ui.presenters.BasePresenter;
import org.amoustakos.utils.android.PackageManagerUtils;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;


public class ActivityListingPresenter<T extends ActivityListingContract.View> extends BasePresenter<T>
											implements ActivityListingContract.Actions {

	private String basePackage;
	private PackageManager pm;


	public ActivityListingPresenter(T view, String basePackage, PackageManager pm) {
		super(view);
		this.basePackage = basePackage;
		this.pm = pm;
	}


	// =========================================================================================
	// Actions
	// =========================================================================================

	@Override
	public void load() {
		Observable.just(true)
			.observeOn(Schedulers.computation())
			.map(bool -> {
				List<ActivityListingModel> models = new ArrayList<>();
				List<Class<? extends BaseActivity>> activities = excludeCurrent(getActivities());

				for (Class<? extends BaseActivity> act : activities) {
						ActivityListingModel model = new ActivityListingModel(
								act,
								act.getName(),
								act.getSuperclass().getName()
						);
						models.add(model);
					}

					Timber.i("Found " + activities.size() + " classes");

					return models;
			})
			.doOnError(Timber::i)
			.onErrorReturn(t -> new ArrayList<>())
			.observeOn(AndroidSchedulers.mainThread())
			.doOnNext(mView::onItemsCollected)
			.subscribe();
	}



	// =========================================================================================
	// Helpers
	// =========================================================================================

	private List<Class<? extends BaseActivity>> getActivities() {
		return PackageManagerUtils.getDefinedActivities(pm, basePackage);
	}

	private List<Class<? extends BaseActivity>> excludeCurrent(List<Class<? extends BaseActivity>> classes) {
		Class<?> current = mView.getClass();
		Integer toRemove = null;

		int count = 0;
		for (Class<? extends BaseActivity> cl : classes) {
			if (current == cl) {
				toRemove = count;
				break;
			}
			count++;
		}

		if (toRemove != null)
			classes.remove(toRemove.intValue());

		return classes;
	}

}

File: app/src/main/java/org/amoustakos/boilerplate/examples/ui/presenters/ActivityListingPresenter.kt
package org.amoustakos.boilerplate.examples.ui.presenters

import android.arch.lifecycle.Lifecycle
import android.content.pm.PackageManager
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.amoustakos.boilerplate.examples.ui.contracts.ActivityListingActions
import org.amoustakos.boilerplate.examples.ui.contracts.ActivityListingView
import org.amoustakos.boilerplate.examples.view.models.ActivityListingModel
import org.amoustakos.boilerplate.ui.activities.BaseActivity
import org.amoustakos.boilerplate.ui.presenters.BasePresenter
import org.amoustakos.utils.android.PackageManagerUtils.definedActivities
import org.amoustakos.utils.android.rx.disposer.LifecycleDisposers
import org.amoustakos.utils.android.rx.disposer.disposeBy
import org.amoustakos.utils.android.rx.disposer.onDestroy
import timber.log.Timber
import java.util.*


class ActivityListingPresenter(
	view: ActivityListingView,
	private val basePackage: String,
	private val pm: PackageManager,
	lifecycle: Lifecycle
): BasePresenter<ActivityListingView>(view, lifecycle), ActivityListingActions {


	private val activities: List<Class<out BaseActivity>>
		get() = definedActivities(pm, basePackage)


	override fun init() {}

	override fun load() {
		Observable.just {}
			.disposeBy(lifecycle.onDestroy)
			.observeOn(Schedulers.computation())
			.map<List<ActivityListingModel>> {
				val models = ArrayList<ActivityListingModel>()
				val activities = excludeCurrent(activities)

				for (act in activities) {
					val model = ActivityListingModel(
                            act,
                            act.simpleName,
                            act.superclass?.name
                    )
					models.add(model)
				}
				models
			}
			.doOnError { Timber.i(it) }
			.onErrorReturn { ArrayList() }
			.observeOn(AndroidSchedulers.mainThread())
			.doOnNext { view().onItemsCollected(it) }
			.subscribe()
	}


	private fun excludeCurrent(classes: List<Class<out BaseActivity>>) =
			classes.filter{ item -> mView.get()?.javaClass != item }

}
